### DISCUSSION TITLE
Debugging and Optimizing Repeated Conditions

### DISCUSSION THREAD

**Discussion by Jonny-T-cyber** [ExperienceLovelace/ha-floorplan#485] [Q&A]
Good morning, everyone,
I’ve written the following code, but it’s not working as expected:

```
- element: screen-message-vassoio-antigoccia
  entities:
    - binary_sensor.power_status
    - sensor.messaggio_hardware
  state_action:
    - action: call-service
      service: floorplan.style_set
      service_data:
        style: >
          ${
            (states['binary_sensor.power_status'].state === "on" && 
            states['sensor.messaggio_hardware'].state === "Vassoio antigoccia estratto") 
              ? "display: inline;" 
              : "display: none;"
          }
    - action: call-service
      service: floorplan.text_set
      service_data:
        element: display-text-row-2
        text: >-
          ${
            (states['binary_sensor.power_status'].state === "on" && 
            states['sensor.messaggio_hardware'].state === "Vassoio antigoccia estratto") 
              ? "Inserire il" 
              : ""
          }
    - action: call-service
      service: floorplan.text_set
      service_data:
        element: display-text-row-3
        text: >-
          ${
            (states['binary_sensor.power_status'].state === "on" && 
            states['sensor.messaggio_hardware'].state === "Vassoio antigoccia estratto") 
              ? "vassoio antigoccia" 
              : ""
          }
    - action: call-service
      service: floorplan.text_set
      service_data:
        element: display-text-row-4
        text: >-
          ${
            (states['binary_sensor.power_status'].state === "on" && 
            states['sensor.messaggio_hardware'].state === "Vassoio antigoccia estratto") 
              ? "e il recipiente dei" 
              : ""
          }
    - action: call-service
      service: floorplan.text_set
      service_data:
        element: display-text-row-5
        text: >-
          ${
            (states['binary_sensor.power_status'].state === "on" && 
            states['sensor.messaggio_hardware'].state === "Vassoio antigoccia estratto") 
              ? "fondi di caffè" 
              : ""
          }
```

I specify that:

1. Only the texts in "**element: display-text-row-2**" and "**element: display-text-row-3**" are displayed.
2. These errors periodically appear in the browser console but they do not appear immediately when I refresh the page:

![image](https://github.com/user-attachments/assets/d3fd9c29-781d-4d8a-a7d7-71dd12b49a95)

Here is the exact text of the errors:
﻿
`0:1 Uncaught (in promise) Error: A listener indicated an asynchronous response by returning true, but the message channel closed before a response was received`

I’d like to ask for your help with the following:

1. Is the provided code free of errors, or does it have any issues?
I want to ensure it’s both syntactically and functionally correct. Also because in the complete listing there are other pieces of code that are written in the same way with different entities.

2. How can I optimize the code to avoid repeating the same condition five times?
I tried storing the result of the condition evaluation in a variable and reusing it later, but I couldn’t get it to work. Is there an effective way to achieve this without introducing errors?

3. What could be the causes of the errors that appear in the console, which seem to occur almost randomly?

Thank you in advance to anyone willing to help

**Comment by exetico** [comment]
Just a quick first input:
Does it work both with `-` and without it, in `style: >`?

For the repeated part, you could use a function, and check what element did trigger the action.

I'm not able to take a closer look right now.

**Reply by Jonny-T-cyber** [reply]
Hi exetico and thanks for the reply

> Does it work both with `-` and without it, in `style: >`?

Yes, it works both with `-` and without it

Regarding the function, it would be a great idea, but...
I'm currently unable to implement it on my own.
If (and when) you have some time, I would be truly grateful if you could spare a few minutes to help me out.

Thank you in advance

**Reply by OzGav** [reply]
I think you need to start small. Get one thing working.

I don't think you can do this:

```
            (states['binary_sensor.power_status'].state === "on" && 
            states['sensor.messaggio_hardware'].state === "Vassoio antigoccia estratto") 
              ? "display: inline;" 
              : "display: none;"
          }
```
I think you should try

```
            (hass.states['binary_sensor.power_status'].state === "on" && 
            hass.states['sensor.messaggio_hardware'].state === "Vassoio antigoccia estratto") 
              ? "display: inline;" 
              : "display: none;"
          }
```

**Reply by exetico** [reply]
How did it go @Jonny-T-cyber ?

**Reply by Jonny-T-cyber** [reply]
> How did it go @Jonny-T-cyber ?

HI @exetico !
I only take up this development now.  I didn't go on.



> For the repeated part, you could use a function, and check what element did trigger the action.

Could you please help me create that function you were talking about and that allows me to avoid repetitions?
Thank you very much

**Reply by Jonny-T-cyber** [reply]
> For the repeated part, you could use a function, and check what element did trigger the action.

Hi @exetico,
please could you help me to implement the function?
I would be immensely grateful
Thanks in advance for your time and your patience

**Reply by exetico** [reply]
Hi @Jonny-T-cyber I'm sorry that I didn't follow up. OzGav did provide one example, but there's actually a few ways of doing this. 

There should not be problems with the source-code itself.

Could you let us know, what you're actually trying to do? I normally mentions xyproblem.info in these kind of requests, as I'm condering why you have that many conditions, for such a small outcome (text_set).
